waspc/src/Wasp/Analyzer/TypeChecker/TypeError.hs

```diff
@@ -43,11 +57,47 @@ data TypeCoerceReason
   | -- | There is no relationship between the types in the coercion
     ReasonUncoercable
   | -- | A coercion of the type contained in a list failed
-    ReasonList TypeError
+    ReasonList e
   | -- | A coercion failed because a dictionary was missing a key
     ReasonDictNoKey String
   | -- | A coercion failed because a dictionary contained an extra key
     ReasonDictExtraKey String
   | -- | A coercion failed because two dictionaries had uncoercable types for a key
-    ReasonDictWrongKeyType String TypeError
+    ReasonDictWrongKeyType String e
   deriving (Eq, Show)
+
+getTypeCoercionErrorMessageAndCtx :: (Type -> TypedExpr -> String) -> TypeCoercionError -> (String, Ctx)
+getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg (TypeCoercionError (WithCtx ctx texpr) t reason) =
+  case reason of
+    ReasonList e ->
+      first (("For list element:\n" ++) . indent 2) $
+        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
+    ReasonDictWrongKeyType key e ->
+      first ((("For dictionary field '" ++ key ++ "':\n") ++) . indent 2) $
+        getTypeCoercionErrorMessageAndCtx getUncoercableTypesMsg e
+    ReasonDictNoKey key -> ("Missing required dictionary field '" ++ key ++ "'", ctx)
+    ReasonDictExtraKey key -> ("Unexpected dictionary field '" ++ key ++ "'", ctx)
+    ReasonDecl -> uncoercableTypesMsgAndCtx
+    ReasonEnum -> uncoercableTypesMsgAndCtx
+    ReasonUncoercable -> uncoercableTypesMsgAndCtx
+  where
+    uncoercableTypesMsgAndCtx = (getUncoercableTypesMsg t texpr, ctx)
+
+getUnificationErrorMessageAndCtx :: TypeCoercionError -> (String, Ctx)
+getUnificationErrorMessageAndCtx = getTypeCoercionErrorMessageAndCtx $
+  \t texpr ->
+    intercalate
+      "\n"
+      [ "Can't mix following types:",
+        concatShortPrefixAndText " - " (show t),
+        concatShortPrefixAndText " - " (show $ exprType texpr)
+      ]
+
+getWeakenErrorMessageAndCtx :: TypeCoercionError -> (String, Ctx)
```

Comment:
```suggestion
getWeakenedErrorMessageAndCtx :: TypeCoercionError -> (String, Ctx)
```
Did you mean this?

Reply:
You could also add a comment potentially about what this does / is used for.

Reply:
Actually no, because we have this operations called "weaken", so when an error occurs, it is called "weakenError".

Reply:
But great idea about the comments, I added those!